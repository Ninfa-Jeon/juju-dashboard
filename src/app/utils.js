import React from "react";

export const generateStatusIcon = (status, count) => {
  let statusClass = status ? `is-${status.toLowerCase()}` : "";
  let countValue = "";
  if (count !== undefined) {
    countValue = ` (${count})`;
  }
  return (
    <span className={"status-icon " + statusClass}>
      {status}
      {countValue}
    </span>
  );
};

export const generateSpanClass = (className, value) => {
  return <span className={className}>{value}</span>;
};
