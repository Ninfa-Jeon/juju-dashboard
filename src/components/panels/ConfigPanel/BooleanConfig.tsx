import { ReactElement, useEffect, useState } from "react";
import classnames from "classnames";

import type { ConfigProps } from "./ConfigPanel";

export default function BooleanConfig({
  config,
  selectedConfig,
  setSelectedConfig,
  setNewValue,
}: ConfigProps): ReactElement {
  const [inputFocused, setInputFocused] = useState(false);

  useEffect(() => {
    if (selectedConfig?.name === config.name) {
      setInputFocused(true);
    } else {
      setInputFocused(false);
    }
  }, [selectedConfig, config]);

  function handleOptionChange(e: any) {
    setNewValue(e.target.name, e.target.value === "true" ? true : false);
  }

  return (
    // XXX How to tell aria to ignore the click but not the element?
    // eslint-disable-next-line
    <div
      className={classnames("config-input", {
        "config-input--focused": inputFocused,
      })}
      onClick={() => setSelectedConfig(config)}
    >
      <h5 className="u-float-left">{config.name}</h5>
      <button className="u-float-right p-button--base">use default</button>
      <div className="row">
        <label className=".p-radio--inline col-2">
          <input
            type="radio"
            className="p-radio__input"
            name={config.name}
            aria-labelledby={config.name}
            defaultChecked={config.value === true}
            value="true"
            onChange={handleOptionChange}
          />
          <span className="p-radio__label">true</span>
        </label>
        <label className="p-radio--inline col-2">
          <input
            type="radio"
            className="p-radio__input"
            name={config.name}
            aria-labelledby={config.name}
            defaultChecked={config.value === false}
            value="false"
            onChange={handleOptionChange}
          />
          <span className="p-radio__label">false</span>
        </label>
      </div>
    </div>
  );
}
