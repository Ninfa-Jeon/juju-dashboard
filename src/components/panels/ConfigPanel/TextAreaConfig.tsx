import { ReactElement, useEffect, useState } from "react";
import classnames from "classnames";

import type { ConfigProps } from "./ConfigPanel";

export default function TextAreaConfig({
  config,
  selectedConfig,
  setSelectedConfig,
  setNewValue,
}: ConfigProps): ReactElement {
  const [inputFocused, setInputFocused] = useState(false);
  const [showUseDefault, setShowUseDefault] = useState(
    config.value !== config.default
  );

  let defaultValue = config.default;
  if (config.default !== config.value) {
    defaultValue = config.value;
  }

  useEffect(() => {
    if (selectedConfig?.name === config.name) {
      setInputFocused(true);
    } else {
      setInputFocused(false);
    }
  }, [selectedConfig, config]);

  return (
    // XXX How to tell aria to ignore the click but not the element?
    // eslint-disable-next-line
    <div
      className={classnames("config-input", {
        "config-input--focused": inputFocused,
      })}
      onClick={() => setSelectedConfig(config)}
    >
      <h5 className="u-float-left">{config.name}</h5>
      <button
        className={classnames("u-float-right p-button--base", {
          "u-hide": !showUseDefault,
        })}
      >
        use default
      </button>
      <textarea
        defaultValue={defaultValue}
        onFocus={() => setSelectedConfig(config)}
        onChange={(e) => {
          setNewValue(config.name, e.target.value);
          if (e.target.value !== config.default) {
            setShowUseDefault(true);
          } else {
            setShowUseDefault(false);
          }
        }}
      ></textarea>
    </div>
  );
}
