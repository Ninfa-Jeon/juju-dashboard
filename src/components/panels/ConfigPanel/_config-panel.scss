@import "vanilla-framework/scss/vanilla";

.config-panel {
  $magic-height: 200px;
  $magic-min-height: calc(100vh - #{$magic-height});

  @media (max-width: $breakpoint-large) {
    grid-template-columns: 1fr !important;
  }

  pre {
    // reset the vanilla pre styling
    background-color: inherit;
    border: none;
    padding: 0;
    // extra styles
    white-space: pre-wrap;
    word-break: keep-all;
  }

  &.row {
    padding: 0;
  }

  .full-size {
    justify-content: center;
    min-height: $magic-min-height;
    text-align: center;
  }

  .config-panel__hide-button {
    @include vf-animation(opacity, brisk, out);
    opacity: 0;
  }

  .config-panel__show-button {
    opacity: 1;
  }

  &__message {
    padding-top: 6rem;

    @media (min-width: $breakpoint-medium) {
      padding-top: 3rem;
    }

    h4 {
      margin-top: 3rem;
    }
  }

  &--center-img {
    margin: 0 auto;
  }

  &__list-header {
    align-items: baseline;
    display: flex;
    height: 2rem;
    justify-content: space-between;
    margin: 2rem 0 1rem;
    padding-left: 0.5rem;

    @media screen and (max-width: $breakpoint-medium) {
      margin-top: 4rem;
    }

    button {
      margin-bottom: 0;
      margin-left: 1rem;
      padding-left: 0.5rem;
      padding-right: 0.5rem;
      white-space: nowrap;
    }

    .entity-name {
      font-size: 1.3rem;
      margin-top: 0.3rem;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }
  }

  &__config-list {
    border-right: 1px solid $color-mid-light;
    padding-top: 1rem;

    @media screen and (min-width: $breakpoint-large) {
      padding-right: 1rem;
    }

    @media screen and (max-width: $breakpoint-large) {
      border-right: none;
    }
  }

  &__list {
    min-height: $magic-min-height;
  }

  &__description {
    margin-top: 2rem;

    @media screen and (max-width: $breakpoint-large) {
      display: none !important;
    }

    &-wrapper {
      position: sticky;
      top: 1rem;
    }
  }

  &__no-description {
    min-height: $magic-min-height;
    padding: 0 5rem;
    position: sticky;
    text-align: center;
    top: 0;
  }

  &__drawer {
    background-color: white;
    border-top: 1px solid $color-mid-light;
    bottom: 0;
    max-height: 130px;
    padding-right: 1.5rem;
    padding-top: 1rem;
    position: sticky;
    text-align: right;

    @media screen and (max-width: $breakpoint-x-small) {
      padding-right: 0;
    }

    button {
      @include vf-animation(opacity, brisk, out);
    }
  }

  &__save-button span {
    padding-left: 1rem;
  }
}

.config-input {
  border-radius: 2px;
  cursor: pointer;
  padding: 0 0.5rem;

  &--focused {
    @include vf-animation(background-color, brisk, out);
    background-color: #ddedfe;
  }

  &--changed {
    background-color: $color-light;
  }

  &.config-input--focused.config-input--changed {
    background-color: #ddedfe;
  }

  .p-icon--plus.config-input--view-description,
  .p-icon--minus.config-input--view-description {
    display: none;
    margin-right: 0.5rem;
    width: 0.875rem;

    @media screen and (max-width: $breakpoint-large) {
      display: inline-block;
    }
  }

  &--description {
    clear: both;
    display: none;
    font-size: 0.875rem;
    margin-bottom: 0.5rem;
    margin-left: 1.3rem;

    @media screen and (max-width: $breakpoint-large) {
      display: block;
    }
  }

  h5 {
    margin-bottom: 0.2rem;
  }

  button {
    color: $color-link;
    margin: 0.3rem 0 0;
    padding: 0;

    &:hover {
      background-color: inherit;
    }
  }

  @media screen and (max-width: $breakpoint-x-small) {
    button {
      width: auto;
    }
  }

  @media screen and (min-width: $breakpoint-large) {
    textarea {
      margin-bottom: 1.2rem !important;
    }
  }

  textarea {
    border: 1px solid #999;
    border-radius: 2px;
    height: 2.3rem;
    margin-bottom: 0.7rem;
    min-height: 2.3rem;
    resize: vertical;
  }
}
