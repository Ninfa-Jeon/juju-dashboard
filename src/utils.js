/**
 * splitStr
 * Split strings on a given character
 *
 * @param {string} str - A string param
 * @param {string} character on which to split string
 * @return {array} Array of strings split
 */
export function splitStr(str, delimiter) {
  return str.split(delimiter);
}
