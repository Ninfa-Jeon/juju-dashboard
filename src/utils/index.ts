export { argPath } from "./argPath";
export { default as getMajorMinorVersion } from "./getMajorMinorVersion";
export { default as getUserName } from "./getUserName";
