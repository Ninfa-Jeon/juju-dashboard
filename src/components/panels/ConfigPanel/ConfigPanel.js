import "./_config-panel.scss";

export default function ConfigPanel() {
  const configSelected = false;
  return (
    <div className="config-panel">
      <div className="row">
        <div className="config-panel__config-list col-6">
          <div className="config-panel__list-header">Icon Mysql</div>
        </div>
        <div className="config-panel__description col-6">
          {configSelected ? (
            <h4>Configuration Description</h4>
          ) : (
            <div className="config-panel__no-description u-vertically-center">
              <div>
                Click on a configuration row to view its related description and
                parameters
              </div>
            </div>
          )}
        </div>
      </div>
    </div>
  );
}
