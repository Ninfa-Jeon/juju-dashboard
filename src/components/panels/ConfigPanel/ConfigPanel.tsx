import { ReactElement, useEffect, useState } from "react";
import { getApplicationConfig, setApplicationConfig } from "juju/index";
import { useStore } from "react-redux";
import type { Store } from "redux";
import classnames from "classnames";
import cloneDeep from "clone-deep";

import Spinner from "@canonical/react-components/dist/components/Spinner";

import { isSet } from "app/utils";
import FadeIn from "animations/FadeIn";

import bulbImage from "static/images/bulb.svg";
import boxImage from "static/images/no-config-params.svg";

import BooleanConfig from "./BooleanConfig";
import TextAreaConfig from "./TextAreaConfig";

import { generateIconImg } from "../../../pages/Models/Details/generators";

import "./_config-panel.scss";

type Props = {
  appName: string;
  charm: string;
  modelUUID: string;
  closePanel: () => void;
};

type ConfigData = {
  name: string;
  default: any;
  description: string;
  source: "default" | "user";
  type: "string" | "int" | "float" | "boolean";
  value: any;
  newValue: any;
};

type Config = {
  [key: string]: ConfigData;
};

export type ConfigProps = {
  config: ConfigData;
  selectedConfig: ConfigData | undefined;
  setSelectedConfig: Function;
  setNewValue: SetNewValue;
};

type SetNewValue = (name: string, value: any) => void;

export default function ConfigPanel({
  appName,
  charm,
  modelUUID,
  closePanel,
}: Props): ReactElement {
  const reduxStore = useStore();
  const [config, setConfig] = useState<Config>({});
  const [selectedConfig, setSelectedConfig] = useState<ConfigData | undefined>(
    undefined
  );
  const [enableSave, setEnableSave] = useState<Boolean>(false);
  const [showResetAll, setShowResetAll] = useState<Boolean>(false);
  const [isLoading, setIsLoading] = useState<Boolean>(true);
  const [savingConfig, setSavingConfig] = useState<Boolean>(false);

  useEffect(() => {
    setIsLoading(true);
    getConfig(
      modelUUID,
      appName,
      reduxStore,
      setIsLoading,
      setConfig,
      checkAllDefaults
    );
  }, [appName, modelUUID, reduxStore]);

  function setNewValue(name: string, value: any) {
    const newConfig = cloneDeep(config);
    newConfig[name].newValue = value;
    if (newConfig[name].newValue === newConfig[name].value) {
      delete newConfig[name].newValue;
    }
    setConfig(newConfig);
    checkEnableSave(newConfig);
    checkAllDefaults(newConfig);
  }

  function checkAllDefaults(config: Config) {
    const shouldShow = Object.keys(config).some((key) => {
      const cfg = config[key];
      if (isSet(cfg.newValue)) {
        if (cfg.newValue === cfg.default) {
          return false;
        } else if (cfg.newValue !== cfg.default) {
          return true;
        }
      } else if (cfg.value !== cfg.default) {
        return true;
      }
      return false;
    });
    setShowResetAll(shouldShow);
  }

  function allFieldsToDefault() {
    const newConfig = cloneDeep(config);
    Object.keys(newConfig).forEach((key) => {
      const cfg = newConfig[key];
      if (cfg.value !== cfg.default) {
        cfg.newValue = cfg.default;
      } else if (cfg.newValue !== cfg.default) {
        cfg.newValue = cfg.default;
      }
    });
    setConfig(newConfig);
    checkAllDefaults(newConfig);
    checkEnableSave(newConfig);
  }

  function checkEnableSave(newConfig: Config) {
    const fieldChanged = Object.keys(newConfig).some(
      (key) =>
        isSet(newConfig[key].newValue) &&
        newConfig[key].newValue !== newConfig[key].value
    );
    setEnableSave(fieldChanged);
  }

  async function handleSubmit() {
    setSavingConfig(true);
    const error = await setApplicationConfig(
      modelUUID,
      appName,
      config,
      reduxStore.getState()
    );
    // It returns an empty object if it's successful.
    if (typeof error === "string") {
      // XXX Surface this to the user.
      console.error("error setting config", error);
    }
    await getConfig(
      modelUUID,
      appName,
      reduxStore,
      setIsLoading,
      setConfig,
      checkAllDefaults
    );
    setSavingConfig(false);
    setEnableSave(false);
  }

  return (
    <div className="config-panel">
      {isLoading ? (
        <div className="full-size u-vertically-center">
          <Spinner />
        </div>
      ) : !isLoading && (!config || Object.keys(config).length === 0) ? (
        <FadeIn isActive={true}>
          <div className="full-size u-align-center">
            <NoConfigMessage />
          </div>
        </FadeIn>
      ) : (
        <FadeIn isActive={true} className="config-content row">
          <div className="config-panel__config-list col-6">
            <div className="config-panel__list-header">
              <div className="entity-name">
                {generateIconImg(appName, charm)} {appName}
              </div>
              <div>
                <button
                  className={classnames(
                    "u-button-neutral config-panel__hide-button",
                    {
                      "config-panel__show-button": showResetAll,
                    }
                  )}
                  onClick={allFieldsToDefault}
                >
                  Reset all values
                </button>
              </div>
            </div>

            <div className="config-panel__list">
              {generateConfigElementList(
                config,
                selectedConfig,
                setSelectedConfig,
                setNewValue
              )}
            </div>
            <div className="config-panel__drawer">
              <button className="p-button--neutral" onClick={closePanel}>
                Cancel
              </button>
              <button
                className={classnames(
                  "p-button--positive config-panel__save-button",
                  {
                    "is-active": savingConfig,
                  }
                )}
                onClick={handleSubmit}
                disabled={!enableSave}
              >
                {!savingConfig ? (
                  "Save and apply"
                ) : (
                  <>
                    <i className="p-icon--spinner u-animation--spin is-light"></i>
                    <span>Saving&hellip;</span>
                  </>
                )}
              </button>
            </div>
          </div>
          <div className="config-panel__description col-6">
            {selectedConfig ? (
              <FadeIn isActive={true}>
                <div className="config-panel__description-wrapper">
                  <h4>Configuration Description</h4>
                  <h5>{selectedConfig.name}</h5>
                  <pre>{selectedConfig.description}</pre>
                </div>
              </FadeIn>
            ) : (
              <div className="config-panel__no-description u-vertically-center">
                <NoDescriptionMessage />
              </div>
            )}
          </div>
        </FadeIn>
      )}
    </div>
  );
}

function getConfig(
  modelUUID: string,
  appName: string,
  reduxStore: Store,
  setIsLoading: (value: boolean) => void,
  setConfig: (value: Config) => void,
  checkAllDefaults: (value: Config) => void
) {
  return getApplicationConfig(modelUUID, appName, reduxStore.getState()).then(
    (result) => {
      // Add the key to the config object to make for easier use later.
      const config: Config = {};
      Object.keys(result.config).forEach((key) => {
        config[key] = result.config[key];
        config[key].name = key;
      });
      setIsLoading(false);
      setConfig(config);
      checkAllDefaults(config);
    }
  );
}

function generateConfigElementList(
  configs: Config,
  selectedConfig: ConfigData | undefined,
  setSelectedConfig: Function,
  setNewValue: SetNewValue
) {
  const elements = Object.keys(configs).map((key) => {
    const config = configs[key];
    if (config.type === "boolean") {
      return (
        <BooleanConfig
          key={config.name}
          config={config}
          selectedConfig={selectedConfig}
          setSelectedConfig={setSelectedConfig}
          setNewValue={setNewValue}
        />
      );
    } else {
      return (
        <TextAreaConfig
          key={config.name}
          config={config}
          selectedConfig={selectedConfig}
          setSelectedConfig={setSelectedConfig}
          setNewValue={setNewValue}
        />
      );
    }
  });

  return elements;
}

function NoConfigMessage() {
  return (
    <div className="config-panel__message">
      <img src={boxImage} alt="" className="config-panel--center-img" />
      <h4>This application doesn't have any configuration parameters</h4>
    </div>
  );
}

function NoDescriptionMessage() {
  return (
    <div className="config-panel__message">
      <img src={bulbImage} alt="" className="config-panel--center-img" />
      <h4>
        Click on a configuration row to view its related description and
        parameters
      </h4>
    </div>
  );
}
