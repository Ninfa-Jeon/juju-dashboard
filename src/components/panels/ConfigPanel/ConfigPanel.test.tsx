import { mount } from "enzyme";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";
import dataDump from "testing/complete-redux-store-dump";

import { waitForComponentToPaint } from "testing/utils";

import ConfigPanel from "./ConfigPanel";

const getApplicationConfig = () => {
  return Promise.resolve({ config: {} });
};

jest.mock("juju/index", () => ({
  getApplicationConfig: getApplicationConfig,
}));

const mockStore = configureStore([]);

describe("ConfigPanel", () => {
  it("displays a message if the app has no config", async () => {
    const store = mockStore(dataDump);
    const title = <div>title</div>;
    const wrapper = mount(
      <Provider store={store}>
        <ConfigPanel
          appName="easyrsa"
          title={title}
          modelUUID=""
          closePanel={() => {}}
        />
      </Provider>
    );

    await waitForComponentToPaint(wrapper);
    expect(wrapper.find(".config-panel__message")).toMatchSnapshot();
  });
});
