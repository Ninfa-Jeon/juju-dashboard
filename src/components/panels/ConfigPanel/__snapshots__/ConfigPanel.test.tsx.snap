// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigPanel displays a message if the app has no config 1`] = `
<div
  className="config-panel__message"
>
  <img
    alt="3d box"
    className="config-panel--center-img"
    src="no-config-params.svg"
  />
  <h4>
    This application doesn't have any configuration parameters
  </h4>
</div>
`;
